The test AddUndertowListenerOnlineTest.addHttpsConnector_commandSucceeds fails on WildFly 10.0.0.CR5

When moving to WildFly 10.0.0.CR5 (see commit https://github.com/Ladicek/creaper/commit/87c25e7898), the test `AddUndertowListenerOnlineTest.addHttpsConnector_commandSucceeds` starts failing with this error:

```
09:14:31,186 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-3) MSC000001: Failed to start service jboss.undertow.listener.test-listener: org.jboss.msc.service.StartException in service jboss.undertow.listener.test-listener: Failed to start service
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1904)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: WFLYUT0079: No SSL Context available from security realm. Either the realm is not configured for SSL, or the server has not been reloaded since the SSL config was added.
	at org.wildfly.extension.undertow.HttpsListenerService.startListening(HttpsListenerService.java:117)
	at org.wildfly.extension.undertow.ListenerService.start(ListenerService.java:138)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1948)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1881)
	... 3 more

09:14:31,195 ERROR [org.jboss.as.controller.management-operation] (management-handler-thread - 2) WFLYCTL0013: Operation ("add") failed - address: ([
    ("subsystem" => "undertow"),
    ("server" => "default-server"),
    ("https-listener" => "test-listener")
]) - failure description: {"WFLYCTL0080: Failed services" => {"jboss.undertow.listener.test-listener" => "org.jboss.msc.service.StartException in service jboss.undertow.listener.test-listener: Failed to start service
    Caused by: java.lang.IllegalStateException: WFLYUT0079: No SSL Context available from security realm. Either the realm is not configured for SSL, or the server has not been reloaded since the SSL config was added."}}
```

Run this command to observe the problem (using Java 8): `mvn clean test -Dtest=AddUndertowListenerOnlineTest#addHttpsConnector_commandSucceeds -DfailIfNoTests=false`

The purpose of this issue is investigation: is it a problem in the test, or did we find a problem in WildFly? (My guess is that it's a problem with the test, there really is no SSL configuration in `ApplicationRealm`, but I'm far from an expert on this.)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@
 - added `OnlineManagementClient.allowFailures` to avoid exceptions
   when operations are executed from commands and failures are expected
   (e.g. `Operations.exists` or `Operations.removeIfExists`)
+- added `AddMariaDb[XA]DataSource`
 
 ## 0.9.2
 
@@ -46,7 +47,7 @@
 ## 0.8.2
 
 - added commands for messaging (both HornetQ and ActiveMQ Artemis)
-- fixed  `DomainAdministration.start|stopServer`
+- fixed `DomainAdministration.start|stopServer`
 
 ## 0.8.1
 

--- a/README.md
+++ b/README.md
@@ -147,17 +147,17 @@ For WildFly 10, which is based on WildFly Core 2:
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-controller-client</artifactId>
-        <version>2.0.4.Final</version>
+        <version>2.0.5.Final</version>
     </dependency>
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-cli</artifactId>
-        <version>2.0.4.Final</version>
+        <version>2.0.5.Final</version>
     </dependency>
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-patching</artifactId>
-        <version>2.0.4.Final</version>
+        <version>2.0.5.Final</version>
     </dependency>
 
 #### Groovy Libraries

--- a/pom.xml
+++ b/pom.xml
@@ -83,6 +83,7 @@
         <version.org.apache.maven.plugins.maven-shade-plugin>2.3</version.org.apache.maven.plugins.maven-shade-plugin>
         <version.org.apache.maven.plugins.maven-source-plugin>2.4</version.org.apache.maven.plugins.maven-source-plugin>
         <version.org.apache.maven.plugins.maven-surefire-plugin>2.18.1</version.org.apache.maven.plugins.maven-surefire-plugin>
+        <version.org.bouncycastle>1.54</version.org.bouncycastle>
         <version.org.codehaus.groovy.groovy-everything>2.3.7</version.org.codehaus.groovy.groovy-everything>
         <version.org.codehaus.mojo.codenarc-maven-plugin>0.22-1</version.org.codehaus.mojo.codenarc-maven-plugin>
         <version.org.codehaus.mojo.findbugs-maven-plugin>2.5.5</version.org.codehaus.mojo.findbugs-maven-plugin>
@@ -187,6 +188,12 @@
                         <artifactId>hamcrest-core</artifactId>
                     </exclusion>
                 </exclusions>
+            </dependency>
+            <dependency>
+                <groupId>org.bouncycastle</groupId>
+                <artifactId>bcprov-jdk15on</artifactId>
+                <version>${version.org.bouncycastle}</version>
+                <scope>test</scope>
             </dependency>
             <dependency>
                 <groupId>org.hamcrest</groupId>

--- a/testsuite/common/pom.xml
+++ b/testsuite/common/pom.xml
@@ -20,6 +20,11 @@
             <scope>compile</scope>
         </dependency>
         <dependency>
+            <groupId>org.bouncycastle</groupId>
+            <artifactId>bcprov-jdk15on</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
             <groupId>org.jboss.shrinkwrap</groupId>
             <artifactId>shrinkwrap-api</artifactId>
             <scope>compile</scope>

--- a/testsuite/common/src/main/java/org/wildfly/extras/creaper/security/KeyPairAndCertificate.java
+++ b/testsuite/common/src/main/java/org/wildfly/extras/creaper/security/KeyPairAndCertificate.java
@@ -1,0 +1,70 @@
+package org.wildfly.extras.creaper.security;
+
+import org.bouncycastle.x509.X509V3CertificateGenerator;
+
+import javax.security.auth.x500.X500Principal;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.security.GeneralSecurityException;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.SecureRandom;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+
+/**
+ * <b>Only for tests!</b> Contains a cryptographic key pair and a certificate for it.
+ */
+public final class KeyPairAndCertificate {
+    private final PrivateKey privateKey;
+    private final PublicKey publicKey;
+    private final X509Certificate certificate;
+
+    private KeyPairAndCertificate(PrivateKey privateKey, PublicKey publicKey, X509Certificate certificate) {
+        this.privateKey = privateKey;
+        this.publicKey = publicKey;
+        this.certificate = certificate;
+    }
+
+    /**
+     * Generates a new key pair (RSA, 2048 bits) and a self-signed certificate for it (SHA1withRSA). The certificate
+     * will use a distinguished name of the form {@code CN=name} and will be valid for 1 year.
+     */
+    public static KeyPairAndCertificate generateSelfSigned(String name) throws GeneralSecurityException, IOException {
+        KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
+        keyPairGenerator.initialize(2048);
+        KeyPair keyPair = keyPairGenerator.generateKeyPair();
+
+        Date from = new Date();
+        Date to = new Date(from.getTime() + 365L * 24L * 60L * 60L * 1000L);
+        BigInteger serialNumber = new BigInteger(64, new SecureRandom());
+        X500Principal owner = new X500Principal("CN=" + name);
+
+        X509V3CertificateGenerator certificateGenerator = new X509V3CertificateGenerator();
+        certificateGenerator.setIssuerDN(owner);
+        certificateGenerator.setSubjectDN(owner);
+        certificateGenerator.setNotBefore(from);
+        certificateGenerator.setNotAfter(to);
+        certificateGenerator.setSerialNumber(serialNumber);
+        certificateGenerator.setPublicKey(keyPair.getPublic());
+        certificateGenerator.setSignatureAlgorithm("SHA1withRSA");
+        X509Certificate certificate = certificateGenerator.generate(keyPair.getPrivate());
+
+        return new KeyPairAndCertificate(keyPair.getPrivate(), keyPair.getPublic(), certificate);
+    }
+
+    /**
+     * Creates a new keystore that will contain a single entry with given {@code entryAlias} and {@code entryPassword}.
+     * The entry will contain the private key and the certificate.
+     */
+    public KeyStore toKeyStore(String entryAlias, String entryPassword) throws GeneralSecurityException, IOException {
+        KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+        keyStore.load(null);
+        keyStore.setKeyEntry(entryAlias, privateKey, entryPassword.toCharArray(), new Certificate[]{certificate});
+        return keyStore;
+    }
+}

--- a/testsuite/pom.xml
+++ b/testsuite/pom.xml
@@ -25,8 +25,8 @@
         <version.wildfly9>9.0.2.Final</version.wildfly9>
         <version.wildfly9.core>1.0.2.Final</version.wildfly9.core>
         <version.wildfly9.arquillian>1.0.2.Final</version.wildfly9.arquillian>
-        <version.wildfly10>10.0.0.CR4</version.wildfly10>
-        <version.wildfly10.core>2.0.4.Final</version.wildfly10.core>
+        <version.wildfly10>10.0.0.CR5</version.wildfly10>
+        <version.wildfly10.core>2.0.5.Final</version.wildfly10.core>
         <version.wildfly10.arquillian>1.1.0.Alpha1</version.wildfly10.arquillian>
 
         <managementPort.as7>9999</managementPort.as7>

--- a/testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/undertow/AddUndertowListenerOnlineTest.java
+++ b/testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/undertow/AddUndertowListenerOnlineTest.java
@@ -1,6 +1,8 @@
 package org.wildfly.extras.creaper.commands.undertow;
 
 import org.jboss.arquillian.junit.Arquillian;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
 import org.wildfly.extras.creaper.core.ManagementClient;
 import org.wildfly.extras.creaper.core.ManagementVersion;
 import org.wildfly.extras.creaper.core.online.CliException;
@@ -10,6 +12,7 @@ import org.wildfly.extras.creaper.core.online.operations.Address;
 import org.wildfly.extras.creaper.core.online.operations.OperationException;
 import org.wildfly.extras.creaper.core.online.operations.Operations;
 import org.wildfly.extras.creaper.core.online.operations.admin.Administration;
+import org.wildfly.extras.creaper.security.KeyPairAndCertificate;
 import org.wildfly.extras.creaper.test.WildFlyTests;
 import org.junit.After;
 import org.junit.Before;
@@ -17,7 +20,10 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.io.File;
+import java.io.FileOutputStream;
 import java.io.IOException;
+import java.security.KeyStore;
 
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
@@ -31,9 +37,13 @@ import static org.junit.Assume.assumeTrue;
 @RunWith(Arquillian.class)
 public class AddUndertowListenerOnlineTest {
     private static final String TEST_LISTENER_NAME = "test-listener";
+    private static final String TEST_PASSWORD = "p4sSw0rD!";
 
     private static final Address DEFAULT_SERVER_ADDRESS = Address.subsystem("undertow")
             .and("server", UndertowConstants.DEFAULT_SERVER_NAME);
+
+    @Rule
+    public final TemporaryFolder tmp = new TemporaryFolder();
 
     private OnlineManagementClient client;
     private Operations ops;
@@ -69,8 +79,23 @@ public class AddUndertowListenerOnlineTest {
 
     @Test
     public void addHttpsConnector_commandSucceeds() throws Exception {
+        String alias = "creaper";
+        File keystoreFile = tmp.newFile();
+        KeyStore keyStore = KeyPairAndCertificate.generateSelfSigned("Creaper").toKeyStore(alias, TEST_PASSWORD);
+        keyStore.store(new FileOutputStream(keystoreFile), TEST_PASSWORD.toCharArray());
+
+        String realmName = "CreaperRealm";
+
+        client.apply(new AddHttpsSecurityRealm.Builder(realmName)
+                .keystorePath(keystoreFile.getAbsolutePath())
+                .keystorePassword(TEST_PASSWORD)
+                .alias(alias)
+                .truststorePath(keystoreFile.getAbsolutePath())
+                .truststorePassword(TEST_PASSWORD)
+                .build());
+
         client.apply(new AddUndertowListener.HttpsBuilder(TEST_LISTENER_NAME, "ajp")
-                .securityRealm("ApplicationRealm")
+                .securityRealm(realmName)
                 .build());
 
         assertTrue(ops.exists(DEFAULT_SERVER_ADDRESS.and("https-listener", TEST_LISTENER_NAME)));
@@ -81,6 +106,8 @@ public class AddUndertowListenerOnlineTest {
                 .forDefaultServer());
         admin.reloadIfRequired();
         assertFalse(ops.exists(DEFAULT_SERVER_ADDRESS.and("https-listener", TEST_LISTENER_NAME)));
+
+        client.apply(new RemoveHttpsSecurityRealm(realmName));
     }
 
     @Test

